### PR TITLE
SNOW-876346: Upgrade urllib to v3.

### PR DESCRIPTION
**Pull request by sfc-gh-pmotacki**
### Description
Upgrading urllib@2 to urllib@3
- [x] installed dependency changed invoking code
- [x] Fixed timeout errors
- [x] Fixed fetching GZIPed files from GCP
- [x] validating OCSP
- [x] validatiing OCSP via proxy - currently call to ocsp responder does not go via proxy

### Checklist
- [x] Format code according to the existing code style (run `npm run lint:check -- CHANGED_FILES` and fix problems in changed code)
- [x] Create tests which fail without the change (if possible)
- [x] Make all tests (unit and integration) pass (`npm run test:unit` and `npm run test:integration`)
- [x] Extend the README / documentation and ensure is properly displayed (if necessary)
- [x] Provide JIRA issue id (if possible) or GitHub issue id in commit message

### CODE REVIEW

File: lib/agent/check.js [hunk 1]
```diff
@@ -100,6 +100,7 @@ function getResponse(uri, req, cb)
     });
   }
 
+  // OCSP is performed without HTTPS
```
Comment (sfc-gh-pfus) [inline lib/agent/check.js:103]: What do you mean by `without HTTPS`? Maybe `insecure HTTP` is better? And why do we need this comment here?

File: lib/agent/https_ocsp_agent.js [hunk 2]
```diff
@@ -2,53 +2,39 @@
...
-var Util = require('../util');
-var HttpsAgent = require('https').Agent;
-var SocketUtil = require('./socket_util');
+const SocketUtil = require('./socket_util');
+const { Agent } = require('urllib');
+const { buildConnector } = require('undici');
+
+function createConnectFunction (connector, mock = undefined) {
+  return function connect ({ hostname, host, protocol, port, servername, localAddress, httpSocket }, callback) {
+    return connector({ hostname, host, protocol, port, servername, localAddress, httpSocket }, (err, socket) => {
```
Comment (sfc-gh-pfus) [inline lib/agent/https_ocsp_agent.js:11]: Why is it better to have error in the first position and the socket in the second? Shouldn't we swap them?

File: lib/agent/https_ocsp_agent.js [hunk 3]
```diff
@@ -2,53 +2,39 @@
...
-  };
-
-  return agent;
+function HttpsOcspAgent () {
+  const connector = buildConnector({});
+  return new Agent({ connect: createConnectFunction(connector) });
 }
 
-module.exports = HttpsOcspAgent;
+module.exports.createConnectFunction = createConnectFunction;
```
Comment (sfc-gh-pfus) [inline lib/agent/https_ocsp_agent.js:41]: Do we need to export this function?

File: lib/agent/socket_util.js [hunk 4]
```diff
@@ -55,7 +55,7 @@ exports.variables = variables;
  *
  * @returns {Object}
  */
-exports.secureSocket = function (socket, host, agent, mock)
+exports.secureSocket = function ( socket, host, agent, mock)
```
Comment (sfc-gh-pfus) [inline lib/agent/socket_util.js:58]: Huge change!

File: lib/agent/socket_util.js [hunk 5]
```diff
@@ -88,15 +89,15 @@ exports.secureSocket = function (socket, host, agent, mock)
       }
       // use ocsp to make sure the entire certificate chain can be trusted
       const certChain = socket.ssl.getPeerCertificate(true);
+
```
Comment (sfc-gh-pfus) [inline lib/agent/socket_util.js:92]: Needed?

File: lib/agent/socket_util.js [hunk 6]
```diff
@@ -110,13 +111,22 @@ exports.secureSocket = function (socket, host, agent, mock)
     Logger.getInstance().trace('socket reused = %s', socket.isSessionReused());
   };
 
-  // when the socket is secure, perform additional validation
-  socket.on('secure', validate);
+  if (socket.authorized) { // when socket is authorized we can check it now
```
Comment (sfc-gh-pfus) [inline lib/agent/socket_util.js:114]: What does `authorized` mean in terms of socket? And what do we check? Do you mean validation?

File: lib/agent/socket_util.js [hunk 7]
```diff
@@ -110,13 +111,22 @@ exports.secureSocket = function (socket, host, agent, mock)
...
+    socket.cork();
+    validate();
+    return {
+      socket,
+      socketError,
+    };
+  } else {
+    // when the socket is secure, perform additional validation
+    socket.on('secure', validate);
```
Comment (sfc-gh-pfus) [inline lib/agent/socket_util.js:123]: What is the difference between this line and 116?

File: lib/http/base.js [hunk 8]
```diff
@@ -93,27 +96,34 @@ HttpClient.prototype.request = function (options)
...
-      // if a callback has been specified, normalize the
-      // response before passing it to the callback
-      if (Util.isFunction(options.callback))
-      {
-        if (err)
-        {
+    if (Util.isFunction(options.callback)) {
+      urllib.request(requestOptions.url, requestOptions)
+        .then(async function ({ data, res }) {
+          if (res.headers['x-goog-stored-content-encoding'] === 'GZIP') { // Google API sets own header and urllib cannot parse it automatically
```
Comment (sfc-gh-pfus) [inline lib/http/base.js:106]: I believe headers may be a list of values and you compare with single one?

File: test/integration/testConnectionWithOCSP.js [hunk 9]
```diff
@@ -113,9 +109,14 @@ describe('Connection with OCSP test', function ()
...
           {
-            assert.strictEqual(err.code, Errors.codes.ERR_SF_NETWORK_COULD_NOT_CONNECT);
-            assert.strictEqual(err.cause.code, Errors.codes.ERR_OCSP_INVALID_VALIDITY);
-            callback();
+            try {
+              assert.strictEqual(err.code, Errors.codes.ERR_SF_NETWORK_COULD_NOT_CONNECT);
+              assert.strictEqual(err.cause.code, Errors.codes.ERR_OCSP_INVALID_VALIDITY);
+              callback();
+            } catch (err) {
+              done(err);
```
Comment (sfc-gh-pfus) [inline test/integration/testConnectionWithOCSP.js:117]: We don't need to check the error here?

File: lib/agent/https_ocsp_agent.js (hunk 2 above)
Comment (sfc-gh-dprzybysz) [reply lib/agent/https_ocsp_agent.js]: it's a used library callback

File: lib/agent/check.js (hunk 1 above)
Comment (sfc-gh-dprzybysz) [reply lib/agent/check.js:103]: Removed comment

File: lib/agent/https_ocsp_agent.js (hunk 3 above)
Comment (sfc-gh-dprzybysz) [reply lib/agent/https_ocsp_agent.js:65]: yes, it's used for tests to pass mock

File: lib/agent/socket_util.js (hunk 4 above)
Comment (sfc-gh-dprzybysz) [reply lib/agent/socket_util.js:58]: fixed

File: lib/agent/socket_util.js (hunk 6 above)
Comment (sfc-gh-dprzybysz) [reply lib/agent/socket_util.js:114]: rephrased comment, check now

File: lib/agent/socket_util.js (hunk 7 above)
Comment (sfc-gh-dprzybysz) [reply lib/agent/socket_util.js:123]: first if branch performs synchronized validation and else performs validation asynchronously

File: lib/http/base.js (hunk 8 above)
Comment (sfc-gh-dprzybysz) [reply lib/http/base.js:105]: Here we expect it to be only GZIP

File: test/integration/testConnectionWithOCSP.js (hunk 9 above)
Comment (sfc-gh-dprzybysz) [reply test/integration/testConnectionWithOCSP.js:117]: throwing the error is enough - any error should fail the test

File: lib/http/base.js (hunk 8 above)
Comment (sfc-gh-pfus) [reply lib/http/base.js:105]: Yes, but does this comparison compares list of strings with simple string? Is the type of `res.headers['x-goog-stored-content-encoding']` is string or list of strings?

File: test/integration/testConnectionWithOCSP.js (hunk 9 above)
Comment (sfc-gh-pfus) [reply test/integration/testConnectionWithOCSP.js:117]: But we expect something specific right? This test would pass if for instance the error is connection refused, not ocsp error?

Comment (sfc-gh-dprzybysz) [reply test/integration/testConnectionWithOCSP.js:117]: done without any paramter or null or undefined means success, if parameter has a value it is an error cause

File: lib/http/base.js (hunk 8 above)
Comment (sfc-gh-dprzybysz) [reply lib/http/base.js:105]: - it's response, not a content negotiation
- we expect only one content encoding here
https://cloud.google.com/storage/docs/xml-api/reference-headers#xgoogstoredcontentencoding

**Comment by sfc-gh-dprzybysz** [comment]
We resigned from using urllib@3 and switched to axios

**Pull request closed without merging**